File: examples/06-openmp/example4_solution.c
// =================================================================
//
// File: example4.c
// Author: Pedro Perez
// Description: This file contains the code to count the number of
//				even numbers within an array using OpenMP.
//
// =================================================================

#include <stdio.h>
#include <stdlib.h>
#include "utils.h"

#include <stdio.h>
#include <stdlib.h>
#include "utils.h"

#define SIZE 1000000000 //1e9

int count_events(int *array, int size) {
	int i, acum;

	acum = 0;
  #pragma omp parallel for shared(array, size) reduction(+:acum)
	for (i = 0; i < size; i++) {
		if (array[i] % 2 == 0) {
			acum++;
		}
	}
	return acum;
}

int main(int argc, char* argv[]) {
	int i, j, *a, result;
	double ms;

	a = (int *) malloc(sizeof(int) * SIZE);
	fill_array(a, SIZE);
	display_array("a", a);

	printf("Starting...\n");
	ms = 0;
	for (i = 0; i < N; i++) {
		start_timer();

		result = count_events(a, SIZE);

		ms += stop_timer();
	}
	printf("result = %i\n", result);
	printf("avg time = %.5lf ms\n", (ms / N));

	free(a);
	return 0;
}

File: examples/java-threads/Example4_Solution.java
// =================================================================
//
// File: Example4_Solution.java
// Author: Pedro Perez
// Description: This file contains the code to count the number of
//				even numbers within an array using Threads.
//
// Copyright (c) 2020 by Tecnologico de Monterrey.
// All Rights Reserved. May be reproduced for any non-commercial
// purpose.
//
// =================================================================

public class Example4_Solution extends Thread {
	private static final int SIZE = 100_000_000;
	private int array[], start, end;
	private int result;

	public Example4_Solution(int array[], int start, int end) {
		this.array = array;
		this.start = start;
		this.end = end;
		this.result = 0;
	}

	public int getResult() {
		return result;
	}

	public void run() {
		result = 0;
		for (int i = start; i < end; i++) {
			if (array[i] % 2 == 0) {
				result++;
			}
		}
	}

	public static void main(String args[]) {
		long startTime, stopTime;
		int array[], block;
		Example4_Solution threads[];
		double ms;
		long result = 0;

		array = new int[SIZE];
		Utils.fillArray(array);
		Utils.displayArray("array", array);

		block = SIZE / Utils.MAXTHREADS;
		threads = new Example4_Solution[Utils.MAXTHREADS];

		System.out.printf("Starting with %d threads...\n", Utils.MAXTHREADS);
		ms = 0;
		for (int j = 1; j <= Utils.N; j++) {
			for (int i = 0; i < threads.length; i++) {
				if (i != threads.length - 1) {
					threads[i] = new Example4_Solution(array, (i * block), ((i + 1) * block));
				} else {
					threads[i] = new Example4_Solution(array, (i * block), SIZE);
				}
			}

			startTime = System.currentTimeMillis();
			for (int i = 0; i < threads.length; i++) {
				threads[i].start();
			}
			for (int i = 0; i < threads.length; i++) {
				try {
					threads[i].join();
				} catch (InterruptedException e) {
					e.printStackTrace();
				}
			}
			stopTime = System.currentTimeMillis();
			ms +=  (stopTime - startTime);

			if (j == Utils.N) {
				result = 0;
				for (int i = 0; i < threads.length; i++) {
					result += threads[i].getResult();
				}
			}
		}
		System.out.printf("sum = %d\n", result);
		System.out.printf("avg time = %.5f ms\n", (ms / Utils.N));
	}
}
